Generate Client using OpenAPI spec tool, Update Overview, Setup Guide in Package.md and ReadMe (#3)
ballerina-platform/module-ballerinax-hubspot.marketing.campaigns: README.md

## Overview

[HubSpot](https://www.hubspot.com/our-story) is a customer platform with all the software, integrations, and resources users need to connect thier marketing, sales, and customer service.

The hubspot.marketing.campaigns package offers APIs to connect and interact with [Marketing Campaigns API](https://developers.hubspot.com/docs/reference/api/marketing/campaigns) endpoints, specifically based on HubSpot API v3.

> [//]: # (TODO: Add overview mentioning the purpose of the module, supported REST API versions, and other high-level details.)

Reviewer: Let's remove the `TODO` part and use the template for the introduction part.

Author: done

## Setup guide

To use the HubSpot Feedback connector, you must have access to the HubSpot API through a HubSpot developer account and a HubSpot App under it. Therefore you need to register for a developer account at HubSpot if you don't have one already.

### Step 1: Create/Login to a HubSpot Developer Account

If you have an account already, go to the [HubSpot developer portal](https://developers.hubspot.com/)

If you don't have a HubSpot Developer Account you can sign up to a free account [here](https://developers.hubspot.com/get-started)

### Step 2 (Optional): Create a [Developer Test Account](https://developers.hubspot.com/docs/getting-started/account-types#developer-test-accounts) under your account

Within app developer accounts, you can create developer test accounts to test apps and integrations without affecting any real HubSpot data.

> *These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts.*

Reviewer: ```suggestion
> **Note:** These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts.
```

Author: done

   1. Go to Test accounts section from the left sidebar.

> ![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/testAccount.png)

Reviewer: Let's add some alt texts for these images. Check similar places.

```suggestion
![test account](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/testAccount.png)
```

Author: done

> ![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/testAccount.png)

Reviewer: Check all the occurrences. 

```suggestion
![](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/testAccount.png)
```

Author: done

   2. Click on the Create developer test account button on the top right corner.

![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/developmentTestAccount.png)

   3. In the pop-up window, provide a name for the test account and click on the Create button.

![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/createAccount.png)

   4. You will see the newly created test account in the list of test accounts.

![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/testAccountPortal.png)

### Step 3: Create a HubSpot App

>    1. Now navigate to the Apps section from the left sidebar and click on the Create app button on the top right corner.

Reviewer: ```suggestion
   1. Navigate to the Apps section from the left sidebar and click on the Create app button on the top right corner.
```

Author: done

![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/appSection.png)

   2. Provide a public app name and description for your app.

![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/namingApp.png)

### Step 4: Setup Authentication

   1. Move to the Auth tab.

![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/auth.png)

   2. In the Scopes section, add the following scopes for your app using the Add new scopes button.
        - marketing.campaigns.read
        - marketing.campaigns.revenue.read
        - marketing.campaigns.write

![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/marketingScopes.png)

>    3. In the Redirect URL section, add the redirect URL for your app. This is the URL where the user will be redirected after the authentication process. You can use localhost for testing purposes. Then hit the Create App button.

Reviewer: ```suggestion
   3. In the Redirect URL section, add the redirect URL for your app. This is the URL where the user will be redirected after the authentication process. You can use `localhost` for testing purposes. Then click the "Create App" button.
```

Author: done

![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/redirectURL.png)

### Step 5: Get the Client ID and Client Secret

Navigate to the Auth tab and you will see the Client ID and Client Secret for your app. Make sure to save these values.

![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/clientId_secretId.png)

### Step 6: Setup Authentication Flow

Before proceeding with the Quickstart, ensure you have obtained the Access Token using the following steps:

   1. Create an authorization URL using the following format:

      ```
      https://app.hubspot.com/oauth/authorize?client_id=<YOUR_CLIENT_ID>&scope=<YOUR_SCOPES>&redirect_uri=<YOUR_REDIRECT_URI>
      ```

>    Replace the <YOUR_CLIENT_ID>, <YOUR_REDIRECT_URI> and <YOUR_SCOPES> with your specific value.

Reviewer: ```suggestion
   Replace the `<YOUR_CLIENT_ID>`, `<YOUR_REDIRECT_URI>`, and `<YOUR_SCOPES>` with your specific value.
```

Author: done

   2. Paste it in the browser and select your developer test account to intall the app when prompted.

![](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/accountSelect.png)

   3. A code will be displayed in the browser. Copy the code.

>       ```
>       Received code: na1-129d-860c-xxxx-xxxx-xxxxxxxxxxxx
>       ```
> 

Reviewer: We can remove this part.

```suggestion
```

Author: done

>    4. Run the following curl command. Replace the <YOUR_CLIENT_ID>, <YOUR_REDIRECT_URI> and <YOUR_CLIENT_SECRET> with your specific value. Use the code you received in the above step 3 as the CODE below

Reviewer: Let's use the [Oxford comma](https://en.wikipedia.org/wiki/Serial_comma) in documentation.  

```suggestion
   4. Run the following curl command. Replace the `<YOUR_CLIENT_ID>`, `<YOUR_REDIRECT_URI>`, and `<YOUR_CLIENT_SECRET>` with your specific value. Use the code you received in the above step 3 as the `CODE` below
```

Author: Ok. done

   - Linux/macOS
   ```bash
   curl --request POST \
--url https://api.hubapi.com/oauth/v1/token \
--header 'content-type: application/x-www-form-urlencoded' \
--data 'grant_type=authorization_code&code=<CODE>&redirect_uri=<YOUR_REDIRECT_URI>&client_id=<YOUR_CLIENT_ID>&client_secret=<YOUR_CLIENT_SECRET>'
   ```
     
   - Windows
   ```bash
curl --request POST ^
--url https://api.hubapi.com/oauth/v1/token ^
--header 'content-type: application/x-www-form-urlencoded' ^
--data 'grant_type=authorization_code&code=<CODE>&redirect_uri=<YOUR_REDIRECT_URI>&client_id=<YOUR_CLIENT_ID>&client_secret=<YOUR_CLIENT_SECRET>'
   ```

This command will return the access token necessary for API calls.

```json
{
  "token_type": "bearer",
  "refresh_token": "<Refresh Token>",
  "access_token": "<Access Token>",
  "expires_in": 1800
}
```

   5. Store the access token securely for use in your application.

> 
> 
> [//]: # (TODO: Add detailed steps to obtain credentials and configure the module.)

Reviewer: ```suggestion
```

Author: removed.

## Quickstart

> To use the HubSpot Marketing Campaigns connector in your Ballerina application, update the .bal file as follows:

Reviewer: ```suggestion
To use the HubSpot Marketing Campaigns connector in your Ballerina application, update the `.bal` file as follows:
```

Author: updated

### Step 1: Import the module

> Import the hubspot.marketing.campaigns module and oauth2 module.

Reviewer: ```suggestion
Import the `ballerinax/hubspot.marketing.campaigns` module and the `ballerina/oauth2` module.
```

Author: updated

```ballerina
import ballerinax/hubspot.marketing.campaigns as hsmcampaigns;
import ballerina/oauth2;
```

### Step 2: Instantiate a new connector

>    1. Create a Config.toml file and, configure the obtained credentials in the above steps as follows:

Reviewer: ```suggestion
   1. Create a `Config.toml` file and, configure the obtained credentials in the above steps as follows:
```

Author: done

```toml
 clientId = <Client Id>
 clientSecret = <Client Secret>
 refreshToken = <Refresh Token>
```

>    2. Instantiate a hsmcampaigns:ConnectionConfig with the obtained credentials and initialize the connector with it.

Reviewer: ```suggestion
   2. Instantiate a `hsmcampaigns:ConnectionConfig` with the obtained credentials and initialize the connector with it.
```

Author: done

```ballerina
configurable string clientId = ?;
configurable string clientSecret = ?;
configurable string refreshToken = ?;

final hsmcampaigns:ConnectionConfig hsmcampaignsConfig = {
    auth : {
        clientId,
        clientSecret,
        refreshToken,
        credentialBearer: oauth2:POST_BODY_BEARER
    }
};

> final hsmcampaigns:Client hsmcampaignsClient = check new (hsmcampaignsConfig, "https://api.hubapi.com");

Reviewer: We can omit the URL here.

```suggestion
final hsmcampaigns:Client hsmcampaignsClient = check new (hsmcampaignsConfig);
```

Author: done

```

### Step 3: Invoke the connector operation

Now, utilize the available connector operations. A sample usecase is shown below.

Create a Marketing Event

```ballerina
public function main() returns error? {
    hsmcampaigns:CollectionResponseWithTotalPublicCampaignForwardPaging campaigns = check baseClient->/marketing/v3/campaigns.get();
}
```

> [//]: # (TODO: Add a quickstart guide to demonstrate a basic functionality of the module, including sample code snippets.)

Reviewer: Let's check and remove all the `TODO` sections.

Author: Removed all

## Overview

[HubSpot](https://www.hubspot.com/our-story) is an AI-powered customer relationship management (CRM) platform. 

> The ballerinax/module-ballerinax-hubspot.marketing.campaigns offers APIs to connect and interact with the [Hubspot Marketing Campaings API](https://github.com/HubSpot/HubSpot-public-api-spec-collection/blob/main/PublicApiSpecs/Marketing/Campaigns%20Public%20Api/Rollouts/177944/v3/campaignsPublicApi.json) endpoints, specifically based on the [Hubspot Marketing Campaings API Documentation](https://developers.hubspot.com/docs/reference/api/marketing/campaigns)

Reviewer: ```suggestion
The `ballerinax/module-ballerinax-hubspot.marketing.campaigns` connector offers APIs to connect and interact with the [Hubspot Marketing Campaigns API](https://developers.hubspot.com/docs/guides/api/marketing/campaigns) endpoints, specifically based on the [HubSpot REST API](https://developers.hubspot.com/docs/reference/api/overview).
```

## Setup guide

> To use the HubSpot Feedback connector, you must have access to the HubSpot API through a HubSpot developer account and a HubSpot App under it. Therefore you need to register for a developer account at HubSpot if you don't have one already.

Reviewer: ```suggestion
To use the HubSpot Marketing Campaigns connector, you must have access to the HubSpot API through a HubSpot developer account and a HubSpot App under it. Therefore you need to register for a developer account at HubSpot if you don't have one already.
```

Author: done

### Step 1: Create/Login to a HubSpot Developer Account

If you have an account already, go to the [HubSpot developer portal](https://developers.hubspot.com/)

If you don't have a HubSpot Developer Account you can sign up to a free account [here](https://developers.hubspot.com/get-started)

### Step 2 (Optional): Create a [Developer Test Account](https://developers.hubspot.com/docs/getting-started/account-types#developer-test-accounts) under your account

> Within app developer accounts, you can create developer test accounts to test apps and integrations without affecting any real HubSpot data.

Reviewer: ```suggestion
### Step 2 (Optional): Create a Developer Test Account

Within app developer accounts, you can create a [developer test account](https://developers.hubspot.com/docs/getting-started/account-types#developer-test-accounts) under your account to test apps and integrations without affecting any real HubSpot data.
```

> **Note**: These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts.

   1. Go to Test accounts section from the left sidebar.

> ![Test Account](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/testAccount.png)

Reviewer: ```suggestion
![Test Account](docs/Resources/testAccount.png)
```

   2. Click on the Create developer test account button on the top right corner.

> ![Development Test Account](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/developmentTestAccount.png)

Reviewer: ```suggestion
![Development Test Account](docs/Resources/developmentTestAccount.png)
```

   3. In the pop-up window, provide a name for the test account and click on the Create button.

> ![Create Account](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/createAccount.png)

Reviewer: ```suggestion
![Create Account](docs/Resources/createAccount.png)
```

   4. You will see the newly created test account in the list of test accounts.

> ![Test Account Portal](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/testAccountPortal.png)

Reviewer: ```suggestion
![Test Account Portal](docs/Resources/testAccountPortal.png)
```

### Step 3: Create a HubSpot App

   1. Navigate to the Apps section from the left sidebar and click on the Create app button on the top right corner.

> ![App Section](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/appSection.png)

Reviewer: ```suggestion
![App Section](docs/Resources/appSection.png)
```

   2. Provide a public app name and description for your app.

> ![Naming the App](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/namingApp.png)

Reviewer: ```suggestion
![Naming the App](docs/Resources/namingApp.png)
```

### Step 4: Setup Authentication

   1. Move to the Auth tab.

> ![Auth Section](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/auth.png)

Reviewer: ```suggestion
![Auth Section](docs/Resources/auth.png)
```

   2. In the Scopes section, add the following scopes for your app using the Add new scopes button.
        - `marketing.campaigns.read`
        - `marketing.campaigns.revenue.read`
        - `marketing.campaigns.write`

> ![Marketing Scopes](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/marketingScopes.png)

Reviewer: ```suggestion
![Marketing Scopes](docs/Resources/marketingScopes.png)
```

   3. In the Redirect URL section, add the redirect URL for your app. This is the URL where the user will be redirected after the authentication process. You can use `localhost` for testing purposes. Then click the "Create App" button.

> ![Redirect URL](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/redirectURL.png)

Reviewer: ```suggestion
![Redirect URL](docs/Resources/redirectURL.png)
```

### Step 5: Get the Client ID and Client Secret

Navigate to the Auth tab and you will see the `Client ID` and `Client Secret` for your app. Make sure to save these values.

> ![Client ID and Client Secret](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/blob/fb6603714f9b563b74775579577749ad62726576/docs/Resources/clientId_secretId.png)

Reviewer: ```suggestion
![Client ID and Client Secret](docs/Resources/clientId_secretId.png)
```

import ballerina/oauth2;
```

### Step 2: Instantiate a new connector

   1. Create a `Config.toml` file and, configure the obtained credentials in the above steps as follows:

```toml
 clientId = <Client Id>
 clientSecret = <Client Secret>
 refreshToken = <Refresh Token>
```

   2. Instantiate a `hsmcampaigns:ConnectionConfig` with the obtained credentials and initialize the connector with it.

```ballerina
configurable string clientId = ?;
configurable string clientSecret = ?;
configurable string refreshToken = ?;

final hsmcampaigns:ConnectionConfig hsmcampaignsConfig = {
    auth : {
        clientId,
        clientSecret,
        refreshToken,
        credentialBearer: oauth2:POST_BODY_BEARER
    }
};

final hsmcampaigns:Client hsmcampaignsClient = check new (hsmcampaignsConfig);
```

### Step 3: Invoke the connector operation

Now, utilize the available connector operations. A sample usecase is shown below.

Retrieve a Marketing Campaign

```ballerina
public function main() returns error? {
    hsmcampaigns:CollectionResponseWithTotalPublicCampaignForwardPaging campaigns = check baseClient->/marketing/v3/campaigns.get();
}
```

## Examples

> The `HubSpot Marketing Campaigns ` connector provides practical examples illustrating usage in various scenarios. Explore these [examples](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/tree/b1251412885d8b083c0df636bfeb85bebc4aeff3/examples), covering the following use cases:

Reviewer: ```suggestion
The `HubSpot Marketing Campaigns ` connector provides practical examples illustrating usage in various scenarios. Explore these [examples](https://github.com/ballerina-platform/module-ballerinax-hubspot.marketing.campaigns/tree/main/examples), covering the following use cases:
```

Author: done

> The `HubSpot Marketing Campaigns ` connector provides practical examples illustrating usage in various scenarios. Explore these [examples](https://github.com/SamudraUduwaka/module-ballerinax-hubspot.marketing.campaigns/tree/b1251412885d8b083c0df636bfeb85bebc4aeff3/examples), covering the following use cases:

Reviewer: Let's include the use case as a list item here. Check the reference connector repo for example.

Author: done
